### PR TITLE
can now normalize iterables keyed with their id

### PR DESCRIPTION
**Pull request by adjohu**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -89,19 +89,19 @@ function visitEntity(entity, entitySchema, bag, options) {
...
 
-function visit(obj, schema, bag, options) {
+function visit(obj, schema, bag, options, collectionKey) {
   if (!isObject(obj) || !isObject(schema)) {
     return obj;
   }
 
   if (schema instanceof EntitySchema) {
-    return visitEntity(obj, schema, bag, options);
+    return visitEntity(...arguments);
```
Comment (paularmstrong) [inline src/index.js]: Can we avoid relying on the `arguments` object in these lines and just add `collectionKey` to the original set?

Comment (adjohu) [inline src/index.js]: could do, just seemed more verbose

Comment (adjohu) [inline src/index.js]: updated @paularmstrong

**Comment by paularmstrong** [comment]
:+1: Thanks @adjohu. Do you want to add info to the docs? (I should have asked before) Otherwise I can do it after I merge.

**Comment by unindented** [comment]
@adjohu could you rebase your PR and add the info to the docs, so that we can get it merged?

**Comment by adjohu** [comment]
Sure.

**Comment by adjohu** [comment]
Ready to merge. @unindented @paularmstrong @gaearon

**Comment by unindented** [comment]
Could you squash the commits and give a meaningful commit message? Just so that the git history is easier to parse for others.

**Comment by adjohu** [comment]
@unindented done

File: README.md [hunk 2]
```diff
@@ -244,11 +244,22 @@ const article = new Schema('articles');
...
 const article = new Schema('articles', { idAttribute: 'slug' });
 
-// Or you can specify a function to infer it
-function generateSlug(entity) { /* ... */ }
+// Or you can specify a function to infer it.
+function generateSlug(entity, collectionKey) { /* ... */ }
 const article = new Schema('articles', { idAttribute: generateSlug });
 ```
 
+If using an function to infer the `idAttribute`, the function is passed both the entity and it's collection key.
```
Comment (unindented) [inline README.md]: Small typo: "If using *a* function"

Comment (unindented) [inline README.md]: Another small typo: "and its collection"

File: README.md [hunk 3]
```diff
@@ -244,11 +244,22 @@ const article = new Schema('articles');
...
+This is helpful in occasions where your data is keyed by id, but doesn't contain id inside the entity. For example:
+
+```javascript
+// {
+//   2: {
+//     name: 'Adam'
+//   } 
+// }
```
Comment (unindented) [inline README.md]: Why comment out the lines? It prevents highlighting...

File: test/index.js [hunk 4]
```diff
@@ -1793,6 +1793,42 @@ describe('normalizr', function () {
...
+      }
+    };
+
+    Object.freeze(input);
+
+    normalize(input, valuesOf(user)).should.eql({
+      result: {
+        1: '1',
+        4: '4'
+      },
```
Comment (unindented) [inline test/index.js]: Does it make sense to have that as a result, instead of `[1, 4]`?

Comment (adjohu) [inline test/index.js]: I agree, but it would take a lot more refactoring to get that result. Seems outside of the scope of the change.

**Comment by unindented** [comment]
I'll let @gaearon and @paularmstrong chime in.

File: README.md [hunk 5]
```diff
@@ -244,11 +244,22 @@ const article = new Schema('articles');
 // You can use a custom id attribute
 const article = new Schema('articles', { idAttribute: 'slug' });
 
-// Or you can specify a function to infer it
-function generateSlug(entity) { /* ... */ }
+// Or you can specify a function to infer it.
+function generateSlug(entity, collectionKey) { /* ... */ }
```
Comment (paularmstrong) [inline README.md]: `key` or `entityKey` would make more sense here.

Even better, the input isn't `entity` (yet), it's the object provided before normalization, so I almost feel like `value`, `key` would make more sense.

File: test/index.js (hunk 4 above)
Comment (paularmstrong) [inline test/index.js]: I'd like to see the result be `[1, 4]` as well, since an array is the typical result format. What work is required for that?

File: README.md (hunk 5 above)
Comment (adjohu) [inline README.md]: value, key makes sense to me too.

**Comment by matt-d-rat** [comment]
It just so happens I have this exact use case now. Looking forward to this being merged 👍

**Comment by adjohu** [comment]
@unindented @gaearon @paularmstrong @matt-d-rat I've updated the PR.

**Comment by adjohu** [comment]
Any reasons not to merge? @unindented

**Comment by paularmstrong** [comment]
👍 for me. I'd like to have @gaearon take care of merging and cutting a release, since this updates the API and docs slightly.

**Comment by adjohu** [comment]
any news?

File: src/index.js [hunk 6]
```diff
@@ -85,28 +85,35 @@ function visitEntity(entity, entitySchema, bag, options) {
...
   } else if (schema instanceof UnionSchema) {
     return visitUnion(obj, schema, bag, options);
   } else {
-    return visitObject(obj, schema, bag, options);
+    return visitObject(obj, schema, bag, options, collectionKey);
   }
 }
 
+function normalizeResult(result) {
+  if (isObject(result) && isEqual(Object.keys(result), Object.values(result))) {
```
Comment (paularmstrong) [inline src/index.js]: I tried to merge this PR and run tests, but it fails a bunch of tests on this line. Can you rebase and fix issues?

**Comment by kristojorg** [comment]
Any updates here?

File: src/index.js (hunk 6 above)
Comment (timhwang21) [inline src/index.js]: The only problem here is `Object.values()`. I can fix this up.

**Comment by paularmstrong** [comment]
@kristojorg The creator gave up and closed their PR. If you'd like to resurrect and reopen a new PR with this, I'd be happy to take another look.

**Comment by timhwang21** [comment]
@paularmstrong doing that now!
